Userland/entr.cpp

```diff
@@ -0,0 +1,203 @@
+        auto line = file->read_line(PATH_MAX);
+        if (line.is_null())
+            break;
+
+        auto line_view = StringView { line.data(), line.size() };
+        String filename = line_view.trim_whitespace(TrimMode::Right);
+        auto real_path = Core::File::real_path_for(filename);
+
+        if (verbose)
+            warnln("[path] {} is {}", filename, real_path);
+
+        auto watch_fd = watch_file(real_path.characters(), real_path.length());
+        if (watch_fd < 0) {
+            perror("watch_file");
+            return 0;
+        }
+        fcntl(watch_fd, F_SETFD, FD_CLOEXEC);
+        auto notifier = Core::Notifier::construct(watch_fd, Core::Notifier::Event::Read);
+        if (verbose)
+            warnln("[watch_file] watching {} with fd={}", filename, watch_fd);
+
+        notifier->on_ready_to_read = [notifier, verbose, clear, force_restart, path = real_path, &executable_name, str_arguments = prepare_arguments(arguments, filename, real_path), m_old_pid = Optional<pid_t> {}] {
+            auto& mutable_old_pid = const_cast<Optional<pid_t>&>(m_old_pid);
+            if (clear)
+                out("\x1b[H\x1b[2J\x1b[3J");
+            char buffer[32];
+            int rc = read(notifier->fd(), buffer, sizeof(buffer));
+            ASSERT(rc >= 0);
+
+            if (verbose)
+                warnln("[notify] Detected change in {}", path);
+
+            if (verbose) {
+                StringBuilder builder;
+                builder.join(' ', str_arguments);
+                warnln("[exec] exec {} with ({})", executable_name, builder.string_view());
+            }
+
+            if (m_old_pid.has_value()) {
+                if (force_restart) {
+                    if (verbose)
+                        warnln("[restart] Killing previous instance for {} with pid={}", path, m_old_pid.value());
+                    kill(m_old_pid.value(), SIGTERM);
+                }
+                int wstatus = 0;
+                waitpid(m_old_pid.value(), &wstatus, 0);
+                mutable_old_pid.clear();
+            }
+
+            Vector<const char*> argv;
+            argv.append(executable_name);
+            for (auto arg : str_arguments)
+                argv.append(arg.characters());
+            argv.append(nullptr);
+
+            posix_spawn_file_actions_t action;
+            posix_spawn_file_actions_init(&action);
+            pid_t pid;
+            if ((errno = posix_spawnp(&pid, executable_name, &action, nullptr, const_cast<char**>(argv.data()), environ))) {
+                perror("posix_spawn");
+                exit(1);
+            }
+            posix_spawn_file_actions_destroy(&action);
+            started_job_ids.append(pid);
+            mutable_old_pid = pid;
+            if (verbose)
+                warnln("[exec] created instance for {} with pid={}", path, pid);
+        };
+
+        notifiers.empend(move(notifier));
+    }
+
+    Core::EventLoop::register_signal(SIGINT, [&](auto) {
```

Comment:
Out of interest, why is this needed? Shouldn't the default SIGINT handler do something similar?

Reply:
I guess, does it hurt to be sure?

Reply:
If process teardown does this for you (and if it didn't, wouldn't it be a bug), why do it manually?